packages/server/src/arpa_reporter/db/arpa-subrecipients.js

```diff
@@ -72,21 +60,67 @@ async function getRecipient(id, trns = knex) {
         .then((rows) => rows[0]);
 }
 
-async function findRecipient(uei = null, tin = null, trns = knex) {
+const SUPPORTED_QUERY_FIELD_TYPES = {
+    UEI: 'uei',
+    TIN: 'tin',
+    NAME: 'name',
+};
+
+/**
+ *
+ * @param { SUPPORTED_QUERY_FIELDS_TYPES } fieldType - the field type to search for
+ * @param { string } value - the value to search for
+ * @param { * } trns - knex transaction
+ * @return {Promise<Subrecipient>} The subrecipient object
+ */
+async function findRecipient(fieldType = null, value = null, trns = knex) {
     const tenantId = useTenantId();
     const query = baseQuery(trns).where('arpa_subrecipients.tenant_id', tenantId);
 
-    if (uei) {
-        query.where('arpa_subrecipients.uei', uei);
-    } else if (tin) {
-        query.where('arpa_subrecipients.tin', tin);
+    if (fieldType === 'uei') {
+        query.where('arpa_subrecipients.uei', value);
+    } else if (fieldType === 'tin') {
+        query.where('arpa_subrecipients.tin', value);
+    } else if (fieldType === 'name') {
+        query.where('arpa_subrecipients.name', value);
     } else {
-        return null;
+        throw new Error('Cannot query for recipient without a valid field type');
     }
 
     return query.then((rows) => rows[0]);
 }
 
+async function createRecipient(recipient, trns = knex) {
+    const tenantId = useTenantId();
+    if (!(recipient.uei || recipient.tin || recipient.name)) {
+        throw new Error('recipient row must include a `uei`, `tin`, or `name` field');
+    }
```

Comment:
I like how the `createRecipient()` function's various `if ... throw` blocks make the business logic clear to developers and provide friendly error messages for users; I'm wondering if we could preserve that clarity while making better use of the db-level constraints by checking the cause of a failure _after_ inserting rather than checking ahead of time (which can fail to identify the problem ahead of time in a double-click situation or similar race condition). 

I think something like this might work:
```js
async function createRecipient(recipient, trns = knex) {
    const tenantId = useTenantId();
    try {
        return await trns('arpa_subrecipients')
            .insert({ ...recipient, tenant_id: tenantId })
            .returning('*')
            .then((rows) => rows[0]);
    } catch (e) {
        switch (e.constraint) {
            case 'chk_at_least_one_of_uei_tin_name_not_null':
                throw new Error('recipient row must include a `uei`, `tin`, or `name` field');
            case 'idx_arpa_subrecipients_tenant_id_uei_unique':
                throw new Error('A recipient with this UEI already exists');
            case 'idx_arpa_subrecipients_tenant_id_tin_unique':
                throw new Error('A recipient with this TIN already exists');
            // Other `case 'name_of_constraint_or_index'` statements go here...
            default:
                // Something else
                throw e;
        }
    }
}
```
(See my previous comment regarding the `chk_at_least_one_of_uei_tin_name_not_null` constraint.)

Reply:
This is much cleaner since we don't need to duplicate the validation logic in both layers. Will update the code 👍

Reply:
added in the latest commit